float32 car_x
float32 car_y
float32 ref_x
float32 ref_y